BLPOP Resp command

rigazilla opened the pull request:
https://issues.redhat.com/browse/ISPN-15148

this is the first implementation of a blocking command.
Is still a WIP.

DONE:
- listener registration
- event loop
- listener de-registration 
- timeout management

MISSING:
~~- timeout management~~

TO BE VERIFIED:
- marshalling of event data


TEST:
- tests for sync connection ~~(usually they pass)~~
- tests for async ~~(lpush fails for wrong password ???)~~
- tests for cleanup and exception handling

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java:
```diff
@@ -174,4 +174,22 @@ public CompletionStage<RespRequestHandler> sendSubscriptions(ChannelHandlerConte
       });
    }
+
+      public CompletionStage<RespRequestHandler> dontSendAndBlock(ChannelHandlerContext ctx, CompletionStage<Void> stageToWaitFor,
```
Seems not used.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,269 @@
+      return pollStage.thenCompose(v -> handler.stageToReturn(
+            (v != null && !v.isEmpty())
+                  ? CompletableFuture.completedFuture(v)
+                  : addSubscriber(arguments, lastKeyIdx, handler, ctx),
```
After the subscription, we need to execute a verification with the pool again. There is a small interval after the checking and before adding the listener in which the entries might be created, and we don't receive an event.

rigazilla replied:
I've added some lines for that

jabolina reviewed (commented):
The structure seems good. I added a few comments. While working for #11223, I added more coverage for different value types, which found a small issue with the listeners.

I added a small commit here changing the listener if you're interested https://github.com/rigazilla/infinispan/compare/BLPOP-15148...jabolina:blpop-listener.

karesti commented on server/resp/src/test/java/org/infinispan/server/resp/RespSingleNodeTest.java:
```diff
@@ -421,6 +428,61 @@ public void testUpperLowercase() {
    }
 
+   @Test
+   void testBlpopAsync() throws InterruptedException, ExecutionException, TimeoutException {
```
this is a list command, should be in the lists test file

karesti replied:
that, or a new file for all the blocking commands testing

rigazilla replied:
added a new RespBlockingCommandsTest.java file

karesti commented on server/resp/src/test/java/org/infinispan/server/resp/RespSingleNodeTest.java:
```diff
@@ -782,21 +843,21 @@ public void testClusterNodesSingleNode() {
-      assertThat(redis.type(k(4))).isEqualTo("zset");
-      assertThat(redis.type(k(100))).isEqualTo("none");
-   }
+   // @Test
```
commented test

karesti reviewed (changes requested):
just a small comment so we have tests organized

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:
```diff
@@ -64,6 +64,10 @@ protected void startInternal() {
+            if (configuredValueType == null) {
+               configuredValueType = MediaType.APPLICATION_OCTET_STREAM;
+               builder.encoding().value().mediaType(configuredValueType);
+            }
```
ok, removed

karesti commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,116 @@
+/**
+ * Base class for single node tests.
+ *
+ * @author William Burns
```
will is working hard!

rigazilla commented:
> The structure seems good. I added a few comments. While working for #11223, I added more coverage for different value types, which found a small issue with the listeners.
> 
> I added a small commit here changing the listener if you're interested [rigazilla/infinispan@BLPOP-15148...jabolina:blpop-listener](https://github.com/rigazilla/infinispan/compare/BLPOP-15148...jabolina:blpop-listener).

thanks! merged

rigazilla commented:
@karesti @jabolina I've implemented the timeout, have a look pls

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,146 @@
+      RedisAsyncCommands<String, String> redis = redisConnection.async();
+      var client = createClient(30000, server.getPort());
+      RedisAsyncCommands<String, String> redis1 = client.connect().async();
+      redis1.lpush("keyY", "firstY").get();
```
I am afraid these async ones with `.get()` would just block until the test suite timeout in case of failures. Looking closely, I think only `redis` needs to be async, and `redis1` could be a `RedisCommands` instead. And we can add some timeout to the `.get()` calls.

jabolina replied:
But the comment I added to the test below should also cover this one.

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,146 @@
+         cf.complete(redisBlock.blpop(0, "keyZ"));
+         return null;
+      });
+      // Ensure lpush is after blpop
```
I guess we can only ensure that using the async client and a latch, not the test would fail. Like:

```java
      CountDownLatch latch = new CountDownLatch(1);
      RedisAsyncCommands<String, String> redisBlock = client.connect().async();
      CompletableFuture<KeyValue<String, String>> cf = CompletableFuture.supplyAsync(() -> {
         RedisFuture<KeyValue<String, String>> rf = redisBlock.blpop(0, "keyZ");
         latch.countDown();
         return rf.toCompletableFuture();
      }, testExecutor()).thenCompose(v -> v);
      // Ensure lpush is after blpop
      assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
```

This should cover the above test, too.

rigazilla replied:
nice!  Implemented, thanks!

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,146 @@
+         res = cf.get(3, TimeUnit.SECONDS);
+         assertThat(res.getKey()).isEqualTo("keyZ");
+         assertThat(res.getValue()).isEqualTo("firstZ");
+      } catch (TimeoutException e) {
```
Shouldn't the test fail in this case?

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,146 @@
+         redisBlock.blpop(-1, "keyZ").get();
+      }).cause().isInstanceOf(RedisCommandExecutionException.class)
+            .hasMessageContaining("ERR value is out of range, must be positive");
+      var tpool = Executors.newCachedThreadPool();
```
Nit, you can use `testExecutor()` instead.

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,146 @@
+ * @since 15.0
+ */
+@Test(groups = "functional", testName = "server.resp.RespBlockingCommandsTest")
+public class RespBlockingCommandsTest extends SingleNodeRespBaseTest {
```
I would also like it if we could test with different value types configured.

rigazilla replied:
what do you mean with different value types?

jabolina replied:
Oh, sorry, I could swear I wrote "value media-type" :sweat_smile: Just to check how the listener would behave.

rigazilla replied:
I've added a RespBlockingMediaTypes.java. I've commented the TEXT/PLAIN case since it fails. I don't know if we need it though

jabolina replied:
Let me run it here to see if I spot something.

jabolina replied:
Also fails with Protostream.

jabolina reviewed (commented):
Looks good. I added a few comments related to the tests. You can squash after and think we're good to go.

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingMediaTypesTest.java:
```diff
@@ -0,0 +1,63 @@
+      MediaType[] types = new MediaType[] {
+            MediaType.APPLICATION_OCTET_STREAM,
+            MediaType.APPLICATION_OBJECT,
+            // MediaType.TEXT_PLAIN,
```
I think is also good to add protostream here.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,165 @@
+         }
+      });
+      // Uninstall the listener when got values
+      return pubSubListener.getFuture().thenApply(v -> {
```
Sorry, I didn't notice at first. This would never finish if the `listnerStage` fails to install the listener. I think if the listener fails to install, we can simply return an error to the user. WDYT?

rigazilla replied:
> Sorry, I didn't notice at first. This would never finish if the `listnerStage` fails to install the listener. I think if the listener fails to install, we can simply return an error to the user. WDYT?

sonds good, I added some lines for that

jabolina replied:
I see the try-catch above, only that wouldn't be enough. I think we can use the `CompletionStages.handleAndCompose` here to handle the failures of the listener completable.

jabolina replied:
Yeah, the problem is caused because we don't return the `listnerStage,` only the `pubSubListener.getFuture().`. To test that case we need to use mocks, got something that simulates the problem:

```java
   public void testBlpopFailsInstallingListener() throws Exception {
      RedisCommands<String, String> redis = redisConnection.sync();
      var client = createClient(30000, server.getPort());
      RedisAsyncCommands<String, String> redisAsync = client.connect().async();

      CacheNotifierImpl<?, ?> cni = (CacheNotifierImpl<?, ?>) TestingUtil.extractComponent(cache, CacheNotifier.class);
      CacheNotifierImpl<?, ?> spyCni = spy(cni);
      CountDownLatch latch = new CountDownLatch(1);

      Answer<CompletionStage<Void>> listenerAnswer = ignore -> {
         latch.countDown();
         return CompletableFuture.failedFuture(new RuntimeException("Injected failure"));
      };

      if (simpleCache) doAnswer(listenerAnswer).when(spyCni).addListenerAsync(any(), any(), any());
      else doAnswer(listenerAnswer).when(spyCni).addListenerAsync(any(), any(), any(), any());

      try {
         TestingUtil.replaceComponent(cache, CacheNotifier.class, spyCni, true);

         CompletableFuture<KeyValue<String, String>> blpop = redisAsync.blpop(0, "my-nice-key")
               .toCompletableFuture();
         assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
         redis.rpush("my-nice-key", "v1", "v2", "v3");
         KeyValue<String, String> res = blpop.get(10, TimeUnit.SECONDS);

         assertThat(res.getKey()).isEqualTo("my-nice-key");
         assertThat(res.getValue()).isEqualTo("v1");
      } finally {
         RespTestingUtil.killClient(client);
         TestingUtil.replaceComponent(cache, CacheNotifier.class, cni, true);
      }
   }
```

This test will fail with a timeout exception because the BLPOP operation never completes. We can catch the exception with the `CompletionStages.handleAndCompose` and complete the `pubSubListener.getFuture().completeExceptionally`. So that stageToReturn() automatically handles it.

jabolina replied:
I'll also check later if we are handling this kind of exception from installing listeners in other commands.

rigazilla replied:
@jabolina thanks for the test case!
I partially rewrote the implementation changing to `addListener` from `addListenerAsync` because the async version hangs at the end of `perform()` and I was unable to find the cause. `addListener` is executed in a stage so should be non-blocking. I kept the commit history if we want to rollback to async.
pls have a look

jabolina commented:
@rigazilla Needs rebase.

karesti commented:
@rigazilla is this ready for merge ?

rigazilla commented:
> @rigazilla is this ready for merge ?

I guess so, @jabolina is reviewing

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,195 @@
+      assertThat(redis.blpop(10, "nullkey")).isNull();
+      var tEnd = timeService.wallClockTime();
+      // The following assertion fails when run with mvn test. Works in IDE
+      // assertThat(tEnd-tStart).isGreaterThanOrEqualTo(5_000L);
```
The `timeService` here in the test is manually controlled, since it wasn't updated both `tStart` and `tEnd` have the same value. You could use the System.currentTimeMillis, I think. Also would suggest a smaller timeout to not take too much to complete.

jabolina replied:
But `testBlpopTimeout` bellow should be covering this already? We could remove this one then.

rigazilla replied:
yeah true

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,195 @@
+      RedisAsyncCommands<String, String> redisBlock = client.connect().async();
+      try {
+         var cf = registerBLPOPListener(redisBlock, 0, "keyZ");
+         Thread.sleep(2000);
```
Since the method above ensures the listener is installed, we can remove the sleeps.

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,195 @@
+
+      try {
+         TestingUtil.replaceComponent(cache, CacheNotifier.class, spyCni, true);
+               assertThatThrownBy(() -> redisAsync.blpop(0, "my-nice-key").get() )
```
I guess we might want a timeout on the `.get` here.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,205 @@
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      cache.addListener(pubSubListener, filter, converter);
```
@wburns Could this still block the event loop?

jabolina replied:
With the listener CF as `CompletionStage<Void> listener = cache.addListenerAsync(pubSubListener, filter, converter);`, what I thought originally was:

```java
      CompletionStage<Void> listener = cache.addListenerAsync(pubSubListener, filter, converter);
      // ... Code handling timeout and pooling again...
      return CompletionStages.handleAndCompose(listener, (ignore, t) -> {
         if (t != null) {
            pubSubListener.getFuture().completeExceptionally(t);
         }

         return pubSubListener.future;
      }).thenApply(ret -> {
         if (scheduledTimer != null) scheduledTimer.cancel(true);
         return ret;
      });
```

rigazilla replied:
nice! applied with some changes

wburns replied:
> @wburns Could this still block the event loop?

Just to close this up, yes it definitely would. I would have expected blockhound to detect this when you have > 1 nodes btw.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,183 @@
+            } else {
+               // ... else wait for values with timeout if required
+               pubSubListener.startTimer(timeout);
+               pubSubListener.getFuture().thenApply(ret -> {
+                  pubSubListener.deleteTimer();
+                  return ret;
+               });
```
We want `.whenComplete` here. Otherwise, if the listener future completes exceptionally, the lambda never executes.

But since we only ever complete exceptionally if `t != null`, we can skip all this timeout install.

rigazilla replied:
I reworked a little bit and added a test to check failure in OnEvent

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,183 @@
+               });
+            }
+            return pubSubListener.getFuture().thenApply(ret -> {
+               cache.removeListenerAsync(pubSubListener);
```
TBH, a bit unsure if we need/how to handle an exception for removing. I see we don't want to fail the request at this point, though.

jabolina reviewed (commented):
Just one tiny thing about removing the timeout.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+   public static class PubSubListener {
+      private final Channel channel;
+      ScheduledFuture<?> scheduledTimer;
+      private CompletableFuture<Collection<byte[]>> future = new CompletableFuture<>();
```
```suggestion
      private final CompletableFuture<Collection<byte[]>> future = new CompletableFuture<>();
```

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+               // ... else wait for values with timeout if required
+               pubSubListener.startTimer(timeout);
+            }
+            pubSubFuture.thenRun(() -> {
```
This doesn't run if the future was completed exceptionally which technically can happen inside the listener itself, which means the listener will be stuck until the timer completes (or forever if the timer didn't have a timeout)

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+            }
+            if (v != null) {
+               // poll returns values, complete the listener future ...
+               pubSubFuture.complete(v);
```
It looks like you have a race condition here where you could have a listener complete this future but with the wrong value if I am looking at the code correctly.

Lets say they ask for keys 1-5 and are all null.

Key 1 gets installed
Then listener is installed
Key 2 gets installed and listener sees it and says it has a value from key 2

Technically key 1 should be the one that was popped instead.

rigazilla replied:
I added a future to sync poll and listener: it waits for the poll stage and return it if not null otherwise it waits for listener events

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+               byte[] key = unwrapKey(entryEvent.getKey());
+               var bucket = (ListBucket<byte[]>) entryEvent.getValue();
+               byte[] value = bucket == null ? null : bucket.toDeque().peek();
+               future.complete(Arrays.asList(key, value));
```
Doesn't this need to pop the value it saw? It seems to only just return it and doesn't pop afai can see.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+            eventType) -> vc.fromStorage(newValue);
+      CompletionStage<Void> addListenerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+      final var pubSubFuture = pubSubListener.getFuture();
+      return CompletionStages.handleAndCompose(addListenerStage, (ignore, t) -> {
```
It doesn't appear we need all of the chained Compose calls. All you need to do is just `return pubSubFuture` as far as I can see as the last line.

wburns commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,257 @@
+         redis.lpush("keyZ", "firstZ");
+         var response = cf.get(10, TimeUnit.SECONDS);
+         assertThat(response.getKey()).isEqualTo("keyZ");
+         assertThat(response.getValue()).isEqualTo("firstZ");
```
We need to assert that the value was actually popped here, as I don't think it was per my comment above.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,230 @@
+                        cache.removeListenerAsync(this);
+                     }
+                  } catch (Exception ex) {
+                     System.out.println(ex.toString());
```
Do we want to catch all Exception here? I would say just let the error propagate to the Stage instead.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,230 @@
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               multimapList.pollFirst(key, 1).handle((v, t) -> {
```
This will still be a problem as we could get multiple events in the time before we remove this listener and thus it would cause multiple keys to be popped. I am thinking we need to fire BLPOP again honestly to remove values as necessary and then ensure with the original BLPOP that it doesn't fire an extra one if the second pollAllKeys found something already.

Also there should be some sort of AtomicBoolean to only fire this once.

rigazilla replied:
I see...
How about something like:
- onEvent side: if the second poll is ongoing save the first received key, else perform the pop and unsubscribe?
- on secondpoll side: on complete if a key has been found pop it, else if a key has been saved from onEvent pop it. unsubscribe

rigazilla replied:
@wburns I redesigned the code following the approach hinted above

wburns reviewed (changes requested):
This still has an issue I described before that if the second pollAllKeys found a key that was valid and the listener also saw one would cause an issue.

karesti commented:
@rigazilla anything else blocking here ?

rigazilla commented:
> @rigazilla anything else blocking here ?

latest version needs review

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,256 @@
+      addListenerStage.whenComplete((ignore, t) -> {
+         // If listener fails to install, complete exceptionally pubSubFuture and return
+         if (t != null) {
+            pubSubListener.completeExceptionally(t);
```
Guessing you want to return to not run the rest of the code on exception?

wburns commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,280 @@
+      // If there's a listener ok otherwise
+      // if rf is done an error during listener registration has happend
+      // no need to wait anymore. test will fail
+      eventually(() -> cni.getListeners().stream()
```
While this provides deterministic behavior of registering the listener, we really need additional tests so that we can test what happens when a listener is not yet installed but a poll event finds a value and also what happens if both the poll event and the listener find the same value.

rigazilla replied:
not sure I can write tests like these, let me see

wburns replied:
We also need a test where we have 2 BLPOP listeners waiting on the same key and see what happens if we add one or two elements to that key. For the one case only one listener should complete and the two case they should both have different values.

rigazilla replied:
Tested in:
testBlpopTwoListenersWithValues
testBlpopTwoListenersOneTimeout

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,256 @@
+      ScheduledFuture<?> scheduledTimer;
+      private boolean listenerAdded;
+      private CompletableFuture<Collection<byte[]>> resultFuture;
+      public boolean pollComplete;
```
This variable is not thread safe. We really need tests as I mentioned below to guarantee concurrent behavior as I am really worried there are a lot of gaps here.

rigazilla replied:
Only one thread changes this variable, the one that completes the poll action, I don't think it needs to be thread safe.
The only concurrent issue I can see here is addressed and commented in the code [here](https://github.com/infinispan/infinispan/blob/921467c37bb69f58f09a96a9c88eb448dd23cbfc/server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java#L228-L229).
Please be more specific about the gaps so I can work on

wburns replied:
It doesn't matter if only one thread changes it unfortunately. If matters if more than one thread read and/or write to it. There is no guarantee the second thread will see the update without encountering a happens before event.

I mentioned the gaps before basically concurrent insertions into the polled keys and what happens if the listener does/doesn't see the update but the second poll does or does not. Also what if listener is listening to > 1 key and if more than 1 one is updated concurrently.

Unfortunately, this command is much more complex than any other and has a high level of concurrency related to it.

rigazilla replied:
Yeah, I'm working on extracting the sync logic in a class so to make easy to write specific unit tests

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,194 @@
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
```
Should be possible to do:

```suggestion
      AdvancedCache<Object, Object> cache = handler.typedCache(null);
```

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,288 @@
+   public static class PollListenerSynchronizer {
+      public CountDownLatch eventReceivedLatch;
+      public CountDownLatch pollGotNullResultLatch;
+      private AtomicReference<byte[]> atomicSavedKey;
+      private CompletableFuture<Collection<byte[]>> resultFuture;
```
```suggestion
      private final AtomicReference<byte[]> atomicSavedKey;
      private final CompletableFuture<Collection<byte[]>> resultFuture;
```

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,288 @@
+      PubSubListener pubSubListener = new PubSubListener(handler, cache, listMultimap);
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
```
Yeah, I think this is a bug in the EncoderCache. This is necessary when using a simple cache. Just below the lines you linked, the data conversion objects are not provided if not a `CacheImpl` instance.

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/commands/list/blocking/PollListenerSynchronizerTest.java:
```diff
@@ -0,0 +1,223 @@
+         synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      }).start();
+      l.countDown();
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
```
I guess we can't tell for sure which one completes.

rigazilla replied:
In this case synchronizer should guarantee that poll value is returned

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/commands/list/blocking/PollListenerSynchronizerTest.java:
```diff
@@ -0,0 +1,223 @@
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void concurrentNullPollAndEventReceived(BLPOP.PollListenerSynchronizer synchronizer)
```
With this synchronization here, I think this should be equivalent to the `onEventThenNullPollTest` method.

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,288 @@
+               multimapList.pollFirst(synchronizer.getSavedKey(), 1)
+                     .whenComplete( (eventVal,t3) -> {
+                     if (t3 != null || eventVal == null || eventVal.size() == 0) {
+                        synchronizer.completeExceptionally(
```
good point, adding more tests

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,288 @@
+         this.cache = cache;
+         this.handler = handler;
+         this.synchronizer = new PollListenerSynchronizer();
+         this.synchronizer.operation = () -> {
```
the idea was to make that part configurable to allow unit testing

wburns replied:
We should instead intercept the addition of the listener to the CacheNotifier and then do replacements there instead, which can be done completely in the test.

rigazilla replied:
mmm yeah, let me see

rigazilla replied:
oh, there's no listener in the unit test. I'll extend the synchronizer and override the operation

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,288 @@
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
```
Maybe we're lucky and the fact that poll is execute both before and after the listener installation can save us from handling mod events. I'm adding more tests to check

rigazilla commented on server/resp/pom.xml:
```diff
@@ -4,7 +4,6 @@
    <modelVersion>4.0.0</modelVersion>
    <parent>
       <?m2e execute onConfiguration,onIncremental?>
```
oh, it's a vscode quickfix to map plugin execution in IDE. should not be committed. removing

karesti commented:
rigazilla wants to merge 32 commits !!!!

rigazilla commented:
> rigazilla wants to merge 32 commits !!!!

No worries @karesti , just keeping the history, instead of force push

It will come a day when all this mess will be squashed in a single commit....
but it is NOT this day! :)

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,275 @@
+      private final EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      private final List<byte[]> keys;
+      private final BiConsumer<? super Collection<byte[]>, ? super Throwable> whenCompleteConsumer;
+      private final AtomicBoolean canPollJustEventKey;
```
Since we're talking about the two variables here (status, keyQueue), I think there is no way to keep them in sync without the synchronized blocks. It would be nice to have everything with just the Atomic* variables, though.

rigazilla replied:
current approach is actually based only on `eventCounter`. 
`canPollJustEventKey` is an optimization to avoid a full poll after the first poll is complete, if I'm corrent an out of sync here should result running a full poll more the necessary.
Will suggested this approach using an atomic counter to match requests and executions, imo it sounds more robust that using states

jabolina reviewed (commented):
Other than the `PollListenerSynchronizer`, I think we need more coverage with multiple servers.

rigazilla commented on server/resp/src/test/java/org/infinispan/server/resp/RespTwoNodeTest.java:
```diff
@@ -106,4 +109,15 @@ public void testPfaddMultipleServers() throws Exception {
    }
+
+   @Test
+   public void testBlpop() throws InterruptedException, ExecutionException {
```
yeah we should add distributed tests. atm we have an issue with datacoverter (coming from SUBSCRIBE command):
#11479

jabolina commented on server/resp/src/test/java/org/infinispan/server/resp/RespTwoNodeTest.java:
```diff
@@ -173,4 +177,17 @@ public void unsubscribed(String channel, long count) {
    }
+
+   @Test
+   public void testBlpop() throws InterruptedException, ExecutionException {
```
Should we include some more here? Installing a listener from client1 but pushing values from client2.

rigazilla replied:
ok, added some tests with mixed cases

jabolina reviewed (commented):
Some minor stuff on the tests. I believe we are ready on the BLPOP itself unless @wburns identifies something.

The only pending point is using BLPOP inside a MULTI. But that should be simple.

rigazilla commented on server/resp/pom.xml:
```diff
@@ -87,7 +88,6 @@
             </configuration>
             <executions>
                <execution>
-                  <?m2e execute onConfiguration?>
```
Do we need this? I added it inadvertently in another PR. It's a quick fix that vscode adds for mvn integration.

wburns replied:
If so, I would do it as a separate PR is all.

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+      private final EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      private final List<byte[]> keys;
+      private final BiConsumer<? super Collection<byte[]>, ? super Throwable> whenCompleteConsumer;
+      private final AtomicBoolean canPollJustEventKey;
```
@wburns I tried to apply your hints, I synchronized `onListenerAdded` and `onEvent` together, I'm not sure this was what you meant  . pls have a look

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+    *
+    */
+   public static class PollListenerSynchronizer {
+      private final ConcurrentLinkedQueue<Object> keyQueue;
```
Change this to an ArrayDeque.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+      }
+
+      private void onEvent(byte[] key) {
+         synchronized (this) {
+            boolean emptyQueue;
+            emptyQueue = keyQueue.isEmpty();
+            keyQueue.offer(key);
+            if (emptyQueue) {
+               runPoll();
+            }
+         }
+      }
```
```suggestion
         boolean emptyQueue;
         synchronized (this) {
            emptyQueue = keyQueue.isEmpty();
            keyQueue.offer(key);
         }
         if (emptyQueue) {
            runPoll(key);
         }
      }
```

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+            } else if (v != null && !v.isEmpty()) {
+               resultFuture.complete(v);
+            } else {
+               if (keyQueue.poll() == this) {
+                  canPollJustEventKey = true;
+               }
+               if (!keyQueue.isEmpty()) {
+                  runPoll();
+               }
```
```suggestion
               byte[] key;
               synchronized (this) {
                 if (keyQueue.poll() == this) {
                     canPollJustEventKey = true;
                  }
                  key = keyQueue.peek();
               }

               if (key != null) {
                  runPoll(key);
               }
```

This needs to be protected by the synchronized block as well. Just wrap the outer if in synchronized(this) block should be all you need.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+         };
+      }
+
+      private void runPoll() {
+         if (canPollJustEventKey) {
+            var key = (byte[]) keyQueue.peek();
+            BLPOP.pollKeyValue(multimapList, key).whenComplete(whenCompleteConsumer);
+         } else {
+            BLPOP.pollAllKeys(multimapList, keys).whenComplete(whenCompleteConsumer);
+         }
+      }
```
```suggestion
      private void runPoll(Object key) {
         if (canPollJustEventKey && key != this) {
            BLPOP.pollKeyValue(multimapList, (byte[]) key).whenComplete(whenCompleteConsumer);
         } else {
            BLPOP.pollAllKeys(multimapList, keys).whenComplete(whenCompleteConsumer);
         }
      }
```
This method can be refactored to pass in the key as all callers have access to it and it would be a lot safer imo as then this method doesn't need to access any variables that are protected by synchronized block.

Note we still have to pay attention to canPollJustEventKey though.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,258 @@
+            emptyQueue = keyQueue.isEmpty();
+            keyQueue.offer(this);
+            if (emptyQueue) {
+               BLPOP.pollAllKeys(multimapList, keys).whenComplete((v, t) -> {
```
Move this out of the synchronized block.

rigazilla commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,576 @@
+         var cf = registerBLPOPListener(redisBlock, 3, "key");
+         var cf2 = registerBLPOPListener(redisBlock2, 3, "key");
+         var cf3 = registerBLPOPListener(redisBlock3, 3, "key");
+         var t1 = new Thread(() -> {
+            redis1.lpush("key", "first");
+         });
+         var t2 = new Thread(() -> {
+            try {
+               Thread.sleep(2000);
+            } catch (InterruptedException e) {
+               // TODO Auto-generated catch block
+               e.printStackTrace();
+            }
+            redis2.lpush("key", "second", "third", "fourth");
+         });
```
I guess is just a debug relic. removed and increased TO

rigazilla commented on server/resp/src/test/java/org/infinispan/server/resp/RespBlockingCommandsTest.java:
```diff
@@ -0,0 +1,576 @@
+      }
+   }
+
+   FailingListener failingListener;
```
nice!

rigazilla commented:
@wburns, I applied your hints. also cleaned up tests as per @jabolina suggestions

wburns reviewed (commented):
Synchronizer looks good 👍

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java:
```diff
@@ -0,0 +1,261 @@
+            keyQueue.offer(this);
+         }
+         if (emptyQueue) {
+            BLPOP.pollAllKeys(multimapList, keys).whenComplete((v, t) -> {
+               whenCompleteConsumer.accept(v, t);
+            });
```
```suggestion
            BLPOP.pollAllKeys(multimapList, keys).whenComplete(whenCompleteConsumer);
```

wburns replied:
Maybe we can clean this up when MULTI for BLPOP is also included.

jabolina reviewed (approved):
Replayed CI.

jabolina commented:
In a follow-up, we can handle the case of mixing BLPOP with MULTI/EXEC.

karesti commented:
Thanks @rigazilla @wburns and @jabolina
